javaparser-core/src/main/java/com/github/javaparser/ast/visitor/TreeVisitor.java

```diff
@@ -23,57 +23,145 @@
 
 import com.github.javaparser.ast.Node;
 
-import java.util.ArrayList;
-import java.util.LinkedList;
-import java.util.Queue;
+import java.util.*;
 
 /**
  * Iterate over all the nodes in (a part of) the AST.
  */
 public abstract class TreeVisitor {
 
     public void visitLeavesFirst(Node node) {
-        for (Node child : node.getChildNodes()) {
-            visitLeavesFirst(child);
-        }
-        process(node);
+        visitPostOrder(node);
     }
 
     /**
-     * Performs a pre-order node traversal starting with a given node. When each node is visited,
-     * {@link #process(Node)} is called for further processing.
+     * Performs a pre-order node traversal starting with a given node. When each node is visited, {@link #process(Node)}
+     * is called for further processing.
      *
      * @param node The node at which the traversal begins.
-     *
      * @see <a href="https://en.wikipedia.org/wiki/Pre-order">Pre-order traversal</a>
      */
     public void visitPreOrder(Node node) {
-        process(node);
-        new ArrayList<>(node.getChildNodes()).forEach(this::visitPreOrder);
+        new PreOrderIterator(node).forEachRemaining(this::process);
+    }
+
+    public static class BreadthFirstIterator implements Iterator<Node> {
```

Comment:
@ftomassetti I turned the tree walking algorithms into iterators so you don't have to visit every node when you only wanted a few, like for a `findFirst`.